### PR TITLE
Manage Vault setting on Gravitywell manifeste

### PR DESCRIPTION
**Pull request by leangeder**
Add the management of the Vault section on the cluster Gravitywell manifest:
Two syntax exist:
##### Direct settings to vault cluster
```
Vault:
  Url: "{VAULT URL}"
  Path: "Path of the Vault Authentification Methods"
  Description: "Description of the Vault Authentification Methods"
```

##### Settings via a git repository
```
Vault:
  Path: "Path of the Vault Authentification Methods"
  Description: "Description of the Vault Authentification Methods"
  Repo:
    Url: "{GIT URL}"
    Branch: "{GIT BRANCH }"
    Path: "Path where Gravitywell will set the vault settings"
```

### CODE REVIEW

File: scheduler/applicationprocessor.go [hunk 1]
```diff
@@ -66,7 +67,7 @@ func groupDeploymentsPerNamespace(cluster configuration.ApplicationCluster) map[
 func executeDeployment(deployment configuration.Application, opt configuration.Options, stateCapture *state.Capture, clusterName string, commandFlag configuration.CommandFlag) {
 	log.Debug(fmt.Sprintf("Loading deployment %s\n", deployment.Name))
 
-	remoteVCSRepoName, err := vcs.FetchRepo(deployment.Git, opt)
+	remoteVCSRepoName, err := vcs.FetchRepo(deployment.Git, opt, "master")
```
Comment (Marvalero) [inline scheduler/applicationprocessor.go]: this should be a constant defined on the topc of the file:
```
const application_default_branch = "master"
```

File: scheduler/clusterprocessor.go [hunk 2]
```diff
@@ -78,10 +76,22 @@ func ClusterProcessor(commandFlag configuration.CommandFlag,
...
 		create := func() {
 			for _, cluster := range provider.Clusters {
-				err := runGCPCreate(cmc, ctx, cluster.Cluster)
+				clusterEndpoint, clusterCertCa, err := runGCPCreate(cmc, ctx, cluster.Cluster)
 				if err != nil {
 					color.Red(err.Error())
 				}
+
+				cluster.Cluster.Endpoint = clusterEndpoint
+				cluster.Cluster.CertCa = clusterCertCa
```
Comment (Marvalero) [inline scheduler/clusterprocessor.go]: as far as I understand, Endpoint and CertCa should NEVER be set by the config file. So I think we could fail if this variables have a value already:
```
if cluster.Cluster.Endpoint != "" {
    log.Error("Cluster Endpoint should not be set ==> ignoring its value")
}
```

File: scheduler/clusterprocessor.go [hunk 3]
```diff
@@ -78,10 +76,22 @@ func ClusterProcessor(commandFlag configuration.CommandFlag,
...
+
+				if err := platform.SetK8SContext(cluster.Cluster); err != nil {
+					color.Red(err.Error())
+				}
+
+				if err := vault.SetVaultConfiguration(opt, cluster.Cluster); err != nil {
```
Comment (Marvalero) [inline scheduler/clusterprocessor.go]: so this function is for google provider, but I guess we want to do this for any provider. So I think I would move this function a little bit down in this file, to the swtich command:

```
func afterCreate(cluster configuration.Cluster) {
			if err := platform.SetK8SContext(cluster.Cluster); err != nil {
				color.Red(err.Error())
			}

			if err := vault.SetVaultConfiguration(opt, cluster.Cluster); err != nil {
				color.Red(err.Error())
			}
}
func afterDelete(cluster configuration.Cluster) {
			if err := platform.UnSetK8SContext(cluster.Cluster); err != nil {
				color.Red(err.Error())
			}

			if err := vault.UnSetVaultConfiguration(opt, cluster.Cluster); err != nil {
				color.Red(err.Error())
			}
}
...

		switch commandFlag {
		case configuration.Create:
			create()
			 afterCreate(cluster)
		case configuration.Apply:
			create()
			 afterCreate(cluster)
		case configuration.Replace:
			delete()
			afterDelete(cluster)
			create()
			afterCreate(cluster)
		case configuration.Delete:
			delete()
			afterDelete(cluster)
		}

```

File: vault/vault.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+	"github.com/fatih/color"
+	v1 "k8s.io/api/core/v1"
+	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func UnSetVaultConfiguration(opt configuration.Options, cluster configuration.ProviderCluster) error {
```
Comment (Marvalero) [inline vault/vault.go]: Please, can you refactor this function? it is really big, it is really difficult to read it

File: vault/vault.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+			}
+
+			defer os.RemoveAll(vaultTempDir)
+			color.Green(fmt.Sprintf("Cluster %s unset on git repo %s on branch %s on the folder %s/unset.", cluster.Name, cluster.Vault.Repo.Url, cluster.Vault.Repo.Branch, cluster.Vault.Repo.Path))
+		}
+	}
+	return nil
+}
+
+func SetVaultConfiguration(opt configuration.Options, cluster configuration.ProviderCluster) error {
```
Comment (Marvalero) [inline vault/vault.go]: Can you refactor this one too?  :)

File: platform/kubernetes.go [hunk 6]
```diff
@@ -26,9 +30,105 @@ import (
 	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	//This is required for gcp auth provider scope
 )
 
+// Setcontext update the file ~/.kube/config with the context associate with teh cluster.
+func UnSetK8SContext(GRVCluster configuration.ProviderCluster) error {
```
Comment (Marvalero) [inline platform/kubernetes.go]: gravitywell just uses GOOGLE_APPLICATION_CREDENTIALS to authenticate with kubernetes, so I do not think this function or SetK8SContext() are needed :thinking:

File: scheduler/applicationprocessor.go (hunk 1 above)
Comment (AlexsJones) [reply scheduler/applicationprocessor.go]: That's not necessary ?

**Comment by AlexsJones** [comment]
Overall, give me a few days to merge the new refactor changes, but I like the idea alot

File: scheduler/clusterprocessor.go (hunk 2 above)
Comment (leangeder) [reply scheduler/clusterprocessor.go]: It is not set by the config file. I have update the file `platform/provider/gcp/create.go` due to return the endpoint and the certificate after the create of the cluster.
This two things have to be provide due to be able to login to the kubernete cluster.

**Comment by leangeder** [comment]
@AlexsJones @Marvalero, Here this is the new PR, https://github.com/AlexsJones/gravitywell/pull/22, after the merge of master on my branch.

**Pull request closed without merging**